xdem/coreg.py

```diff
@@ -1063,3 +1081,620 @@ def coregister(reference_raster: Union[str, gu.georaster.Raster], to_be_aligned_
     )
 
     return aligned_raster, error
+
+
+def _transform_to_bounds_and_res(dem: np.ndarray,
+                                 transform: rio.transform.Affine) -> tuple[rio.coords.BoundingBox, float]:
+    bounds = rio.coords.BoundingBox(
+        *rio.transform.array_bounds(dem.shape[0], dem.shape[1], transform=transform))
+    resolution = (bounds.right - bounds.left) / dem.shape[1]
+
+    return bounds, resolution
+
+
+class Coreg:
+    _meta: Optional[dict[str, Any]] = None  # All __init__ functions should instantiate an empty dict.
+    _fit_called = False  # Flag to check if the .fit() method has been called.
+
+    def __init__(self):
+        """This function should have been overwritten by subclassing."""
+        raise ValueError("Coreg class should not be instantiated directly.")
+
+    def fit(self, reference_dem: Union[np.ndarray, np.ma.masked_array],
```

Comment:
Something I wanted to mention, but we should make sure that the functions do not only accept DEMs, but also point clouds. Many people use ICESat/ICESat-2 data as well for coregistration. For most applications, a raster is not needed.
I've added a note in the Project's tab about this.

Reply:
Let's have this as a future improvement! I see this working great with ICP and NuthKaab, but not as easily with dDEM-dependent analyses: Deramp and BiasCorr. I'm sure there's a way, though, and we can explore that in the future!

Reply:
Oh that would work with deramp and biascorr for sure. My own dem_coregistration.py script was able to handle point data in hdf5 format.
All basic functions actually only need a 1D array of (x, y, z) values. Except of course for NuthKaab were we need to calculate the slope/aspect before "destructuring" the raster. But get_hozitonal_shift does not require 2-D data.